File: src/whisper_typer/code_speaker/reminder.py
"""Escalating reminder system for code_speaker TTS notifications."""

import asyncio
import logging
from typing import Optional

logger = logging.getLogger(__name__)


class ReminderManager:
    """Manages repeating TTS reminders until user responds.

    After the initial TTS notification, repeats the same message
    every `interval` seconds until cancelled (by UserPromptSubmit).
    """

    def __init__(self, interval: int = 300):
        self.interval = interval
        self._task: Optional[asyncio.Task] = None
        self._speak_fn = None
        self._reminder_count = 0

    @property
    def is_active(self) -> bool:
        return self._task is not None and not self._task.done()

    @property
    def reminder_count(self) -> int:
        return self._reminder_count

    def start(self, text: str, speak_fn):
        """Start repeating reminders for given text.

        Args:
            text: The text to speak on each reminder.
            speak_fn: Async callable that speaks the text.
        """
        self.cancel()
        self._speak_fn = speak_fn
        self._reminder_count = 0
        self._task = asyncio.ensure_future(self._reminder_loop(text))
        logger.info(f"Reminder started: every {self.interval}s")

    async def _reminder_loop(self, text: str):
        """Loop: sleep → speak → repeat."""
        try:
            while True:
                await asyncio.sleep(self.interval)
                self._reminder_count += 1
                logger.info(f"Reminder #{self._reminder_count}: speaking")
                if self._speak_fn:
                    await self._speak_fn(text)
        except asyncio.CancelledError:
            logger.info(f"Reminder cancelled after {self._reminder_count} reminders")

    def cancel(self):
        """Cancel all pending reminders."""
        if self._task and not self._task.done():
            self._task.cancel()
        self._task = None
        count = self._reminder_count
        self._reminder_count = 0
        return count

File: whisper-typer-rs/src/code_speaker/reminder.rs
//! Periodic TTS reminder manager.
//!
//! Repeats a TTS notification at intervals until cancelled.
//! Used for "Claude is waiting" reminders after task completion.

use std::sync::atomic::{AtomicBool, AtomicU32, Ordering};
use std::sync::{Arc, Mutex};
use std::time::Duration;

use tokio::task::JoinHandle;
use tracing::info;

use super::tts::KokoroTtsEngine;

pub struct ReminderManager {
    interval_secs: u64,
    task: Mutex<Option<JoinHandle<()>>>,
    active: Arc<AtomicBool>,
    count: Arc<AtomicU32>,
}

impl ReminderManager {
    pub fn new(interval: u64) -> Self {
        Self {
            interval_secs: interval,
            task: Mutex::new(None),
            active: Arc::new(AtomicBool::new(false)),
            count: Arc::new(AtomicU32::new(0)),
        }
    }

    pub fn is_active(&self) -> bool {
        self.active.load(Ordering::Relaxed)
    }

    pub fn reminder_count(&self) -> u32 {
        self.count.load(Ordering::Relaxed)
    }

    /// Start repeating reminders. Cancels any existing reminder first.
    pub fn start(&self, text: String, tts: Arc<KokoroTtsEngine>) {
        self.cancel();
        self.active.store(true, Ordering::Relaxed);
        self.count.store(0, Ordering::Relaxed);

        let active = self.active.clone();
        let count = self.count.clone();
        let interval = self.interval_secs;

        let handle = tokio::spawn(async move {
            loop {
                tokio::time::sleep(Duration::from_secs(interval)).await;
                if !active.load(Ordering::Relaxed) {
                    break;
                }
                let n = count.fetch_add(1, Ordering::Relaxed) + 1;
                info!("Reminder #{n}: speaking");
                let _ = tts.speak(&text).await;
            }
        });

        *self.task.lock().unwrap() = Some(handle);
    }

    /// Cancel all pending reminders. Returns count of reminders fired.
    pub fn cancel(&self) -> u32 {
        self.active.store(false, Ordering::Relaxed);
        if let Some(handle) = self.task.lock().unwrap().take() {
            handle.abort();
        }
        self.count.swap(0, Ordering::Relaxed)
    }
}
